ifrs17/Utils/ActivityLogsAndImports.ipynb

```diff
@@ -0,0 +1,302 @@
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public record KeyedDataSetImport : ImportWithKey",
+                "\n{",
+                "\n    public KeyedDataSetImport(DataSetImportOptions options)",
+                "\n    {",
+                "\n        importOptions = options;",
+                "\n        Id = Guid.NewGuid();",
+                "\n    }",
+                "\n}"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public record KeyedStreamImport : ImportWithKey",
+                "\n{",
+                "\n    public KeyedStreamImport(StreamImportOptions options)",
+                "\n    {",
+                "\n        importOptions = options;",
+                "\n        Id = Guid.NewGuid();",
+                "\n    }",
+                "\n}"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public static async Task<ActivityLog> ExecuteWithStoreActivityAsync(this ImportOptionsBuilder builder, ",
+                "\n                                                                                ISessionVariable session, ",
+                "\n                                                                                IDataSource dataSource)",
+                "\n{",
+                "\n    var log = await builder.ExecuteAsync();",
+                "\n    var options = builder.GetImportOptions();",
+                "\n    var activity = new KeyedActivity(log, session);",
+                "\n    switch(options)",
+                "\n    {",
+                "\n        case FileImportOptions fio:",
+                "\n            var keyedFile = await (new KeyedFile(fio)).InitializeImportDataAsync(session);",
+                "\n            activity = activity with {SourceId = keyedFile.Id, ",
+                "\n                                        Category = \"Import from File\"};",
+                "\n            if (keyedFile is KeyedFile kf) await dataSource.UpdateAsync<KeyedFile>(kf.RepeatOnce());",
+                "\n            break;",
+                "\n        case StringImportOptions sgio:",
+                "\n            var keyedStringImport = await (new KeyedStringImport(sgio)).InitializeImportDataAsync(session);",
+                "\n            activity = activity with {SourceId = keyedStringImport.Id, ",
+                "\n                                    Category = \"Import from String\"};",
+                "\n            if (keyedStringImport is KeyedStringImport ksi) await dataSource.UpdateAsync<KeyedStringImport>(ksi.RepeatOnce());",
```

Comment:
Why do you need the if? You construct the variable 2 lines above an the type is known, right?

Reply:
Yep, I moved now the explicit cast 3 rows above, where it belongs and got rid of the if statement (InitializeImportDataAsync returns the abstract class ImportWithKey, so an explicit cast back into KeyedStringImport is needed).